Add 'node to participant mapping' article (#1)
ivanpauno/design: articles/151_node_to_participant_mapping.md

---
layout: default
title: Node to Participant mapping
permalink: articles/node_to_participant_mapping.html

> abstract: This article analyze the performance implications of having a one-to-one Node to Participant mapping and propose an alternative approach.

Reviewer: @ivanpauno consider
```suggestion
abstract: This article analyzes the performance implications of enforcing a one-to-one mapping between ROS nodes and DDS participants, and proposes alternative approaches.
```

>   How Nodes are mapped to DDS participants.

Reviewer: @ivanpauno leftover?

author: '[Ivan Paunovic](https://github.com/ivanpauno)'
published: true
categories: Middleware
---

{:toc}

# {{ page.title }}

<div class="abstract" markdown="1">
{{ page.abstract }}
</div>

Original Author: {{ page.author }}

## Introduction

### What is a `Node`?

> In ROS, a `Node` is an entity used to group other entities.

Reviewer: @ivanpauno hmm I wouldn't say it groups other _entities_, as opposed to _interfaces_.

For example: `Publishers`, `Subscriptions`, `Services`, `Clients`.

> `Nodes` ease organization and code reusage, as they can be composed or launched in different ways.

Reviewer: @ivanpauno consider
```suggestion
`Nodes` ease organization and code reuse, as they can be composed and launched in different ways.
```
instead

### What is a `Domain Participant`?

A `Domain Participant` is a type of DDS entity.
`Participants` also group other entities, like `Publishers`, `Subscribers`, `Data Writters`, `Data Readers`, etc.
But participants do more than that:

- Each `Participant` does discovery by its own.

>   Creating more than one `Participant` increase cpu and network.

Reviewer: @ivanpauno 
```suggestion
  Creating more than one `Participant` increases CPU usage and network IO load.
```

> - Each `Participant` keeps track of the others and of each DDS entity.

Reviewer: @ivanpauno what does `others` mean in this context i.e. let's be explicit.

  Using more than one will duplicate that data.

> - Each `Participant` creates multiple threads for event handling, discovery, etc.

Reviewer: @ivanpauno I'd bind this statement with particular DDS implementations (as it's an implementation detail).

Author: I'm not sure if it's part of the standard or not.
I will rephrase to: `... may create`, and add a cite to `connext` documentation.

Running a listener demo with fastrtps and running `ps -elfT`:
```
0 S ivanpau+ 25754 25754 25749  0  80   0 - 108451 futex_ 16:48 pts/0   00:00:00 /home/ivanpauno/ros2_ws/install/demo_nodes_cpp/lib/demo_nodes_cpp/listener
1 S ivanpau+ 25754 25755 25749  0  80   0 - 108451 futex_ 16:48 pts/0   00:00:00 /home/ivanpauno/ros2_ws/install/demo_nodes_cpp/lib/demo_nodes_cpp/listener
1 S ivanpau+ 25754 25756 25749  0  80   0 - 108451 ep_pol 16:48 pts/0   00:00:00 /home/ivanpauno/ros2_ws/install/demo_nodes_cpp/lib/demo_nodes_cpp/listener
1 S ivanpau+ 25754 25757 25749  0  80   0 - 108451 skb_wa 16:48 pts/0   00:00:00 /home/ivanpauno/ros2_ws/install/demo_nodes_cpp/lib/demo_nodes_cpp/listener
1 S ivanpau+ 25754 25758 25749  0  80   0 - 108451 skb_wa 16:48 pts/0   00:00:00 /home/ivanpauno/ros2_ws/install/demo_nodes_cpp/lib/demo_nodes_cpp/listener
1 S ivanpau+ 25754 25759 25749  0  80   0 - 108451 skb_wa 16:48 pts/0   00:00:00 /home/ivanpauno/ros2_ws/install/demo_nodes_cpp/lib/demo_nodes_cpp/listener
1 S ivanpau+ 25754 25760 25749  0  80   0 - 108451 futex_ 16:48 pts/0   00:00:00 /home/ivanpauno/ros2_ws/install/demo_nodes_cpp/lib/demo_nodes_cpp/listener
```
7 threads! and it's using a single threaded executor (some of them may be from ROS).

Doing the same with `manual_composition` demo creates 19 threads.
The demo compose 4 nodes.
I guess FastRTPS creates 4 nodes per participant, and we always create two background threads independently of the number of Nodes (one is for ctrl-c, the other IDK it may come from fastrtps too).

Composing 20 nodes would create 83 threads.

  Creating more than one will increase cpu and memory usage.

For those reasons, `Participants` are heavyweight.

### Current status

> There is a one to one map between `Nodes` and `DDS Participants`.

Reviewer: @ivanpauno consider `one-to-one mapping` instead.

This simplified the design, as `DDS Participants` provide the same organization that a `Node` needs.
The drawback of this approach, is that performance is deteriorated.

> Furthermore, the number of `Domain participants` is limited to a small number.

Reviewer: @ivanpauno consider 
```suggestion
Furthermore, the maximum number of `Domain participants` is rather small.
```

For example, [RTI connext](https://community.rti.com/kb/what-maximum-number-participants-domain) is limited to 120 participants per domain.

## Proposal

The goal of this proposal is to improve overall performance by avoiding the creation of one `Domain Participant` per `Node`.
API changes will be avoided, if possible.

### What is a participant mapped to?

There are two main alternatives:
- One participant per process.

> - One participant per context.

Reviewer: @ivanpauno it'd be good to define `context` here or above.

The second approach allows more flexibility.
Considering that by default there's only one context per process, it won't lower the performance.

> Moreover, a mechanism for using the same participant in two context could be added.

Reviewer: @ivanpauno consider 
```suggestion
Moreover, a mechanism for re-using the same participant in two separate contexts could be added.
```

### What is a Node now?

> There's not lightweight DDS entity similar to a ROS `Node`, so `Nodes` have to be built from scratch.

Reviewer: @ivanpauno consider 
```suggestion
There's no lightweight DDS equivalent of a ROS `Node`, so these must be implemented on top of it.

A `Node` should be able to:
- Create other entities as `Publishers`, `Subscriptions`, `Services` and `Clients`.
  `Nodes` should own those entities, that is to say, those entity shouldn't outlive a `Node`.
- List all its entities.

> For all the entities, it should be possible to get the `Node` what created them.

Reviewer: @ivanpauno `what` -> `that`

> Each `Participant` should store all the information needed about its nodes, and communicate it to the others.

Reviewer: @ivanpauno `... and communicate it to other Participants`.

### How `Node` information is communicated?

#### Using a topic

The name of all the available `Nodes`, and its `Publishers`, `Subscriptions`, `Services`, `Clients` should be available for every `Participant`.
This information can be communicated using a `topic`.

> That topic will be an implementation detail and hidden to the user.

Reviewer: @ivanpauno what does hidden to the user mean? Hidden by convention (i.e. leading underscore), hidden by not abiding to the ROS convention for DDS topic names (i.e. without an `rt/` prefix) ?

Author: The second one, I will be more explicit.

> A message could be send for:

Reviewer: @ivanpauno `send` -> `sent`

- Each `Node`
- Each `Participant`

> The second option reduces the amount of messages and it does lifetime management easier, as the message will not be more available after the `Participant` is deleted (using the correct QoS policies).

Reviewer: @ivanpauno hmm I'm not sure I follow this. Isn't it that we leverage liveliness QoS settings, regardless of whether the message information is kept by other participants?

Author: Deleted, and added other advantages of the second approach.

In the following, the second option will be considered.

##### State Message

Each `Participant` will send a message representing their state.
A keyed topic could be used for communicating it.

> It's reasonable to organize the data by `Participant`, so the `Participant` guid can be used as the key.

Reviewer: @ivanpauno hmm I'd drop `It's reasonable`, sounds subjective :)

Reviewer: Also, `guid` -> `GUID`

> The rest of the message will be a vector of with the information of each node.

Reviewer: @ivanpauno I'd say `array` or `sequence` instead of `vector`, so that it matches IDL terminology.

That message should contain the `Node` name, and four vectors:
- guid of its `Publishers`
- guid of its `Subscriptions`
- guid of its `Services`
- guid of its `Clients`

Vector bounds: TBD

##### QoS for communicating node information

> Each published message should be available to late subscriptions, and only the last message of each key is needed.

Reviewer: @ivanpauno `late subscriptions` -> `late subscribers`

Reviewer: And `... of each key should be kept.`

For that reason, the QoS of the `Publishers` should be:

- Durability: Transient Local
- History: Keep Last
- History depth: 1
- Reliability: Reliable

Considering that a keyed topic will be used, in which the history depth apply for each key, only one `Publisher` per process will be needed.

The configuration of the `Subscriber` QoS depends on how the data will be accessed later:
- Polled using `Subscriber` read method when needed.
- Listened and organized in a local cache.

The second option allows better organization of this information (e.g.: in hash tables).
In the first case, the QoS of the `Subscriber` should be:

- Durability: Transient Local
- History: Keep Last
- History depth: 1
- Reliability: Reliable

In the second case, durability can be changed to `Volatile`.

#### Using USER_DATA and GROUP_DATA QoSPolicy

Each `Participant` could have in its user data the list of node names that owns.
When this data is changed, each `ParticipantListener` will be notified.

> This is not a good option, as `UserData` is just a sequence of bytes, and communicating a list of node names in it doesn't seem as a reasonable solution.

Reviewer: @ivanpauno I'd be explicit about the rationale of why this isn't a good option.

Similarly to `UserData`, `GroupData` is a available in `Publishers` and `Subscribers`.
These entities only need to communicate the guid of the `Participant` and the `Node` name from which it was created.
This idea can be combined with a topic just publishing the list of `Node` names, without including all the other vectors in the message.
Although, it is more difficult to communicate this information for `Services` and `Clients`, as they use behind the scenes just a `DDS Publisher` and `Subscriber`.

### Other implications

#### Security

In `DDS`, security can be specified at a `Participant` level.

> Currently, as each `Node` was mapped to one `Participant`, we can individually configure its security key and access control policy.

Reviewer: @ivanpauno I'd avoid saying `currently` and rephrase so that this article remains valid in the future.

From a security point of view, only being able to configure it at a `Participant` (or per process) level is enough.
There's not much sense on having different access control policies for `Nodes` in the same process.
As they share the same address space, other vulnerabilities are possible.

##### How to create a new security key?

Before, we were creating a key for each `Node`.
The full name of the node was used for creating it.

If we create one `Participant` per context, we will only need a key for each of them, and not one per `Node`.
There are two alternatives:

- Add the concept of `Context` name (or `Participant` name).
  In this way, the key of each participant could be specified independently.
- Use one key per process.
  All the `Participants` within one process will use the same key (is that possible?).

##### How to specify access policies?

Access control policies could still be specified per `Node` basis.
When a `Participant` is created, it should look at the access control policies of each of its `Nodes` and compose them in a single configuration.

QQ:

> - Is it possible to add more access control policies after creating the `Participant` (e.g.: When later creating a `Node`).

Reviewer: @ivanpauno I'd remove sentence and put it in a comment on the PR for further discussion.

> #### Node Name Uniqueness

Reviewer: @ivanpauno though important, it seems to me that node name uniqueness is not a necessary condition for a one-to-many mapping DDS participants and nodes. If so, I think it makes sense to move that discussion to the right article.

Author: I will left a comment saying that we don't need to address it in the first iteration.
But I think we have to left this section.
The problem is different when enforcing a one-to-one mapping than when not.
So, an explanation that it isn't more difficult is fair.

> We aren't currently enforcing `Node` name uniqueness.

Reviewer: @ivanpauno same about using `currently`

When creating only one `Participant` per `Context`, we can distinguish two cases:
- There is an overlap between the name of two `Nodes` created within the same `Context`.
  This case can be trivially solved.
- There is a collision with the `Node` name created from another `Context`.
  By the nature of discovery, when a collision is detected, it's not possible to know what `Node` was created first without extra information.
  A collision resolution mechanism have to be decided for solving which `Node` continues living.
  A `timestamp` of the `Node` creation published in the state message can help to solve the problem.

If we don't change the `Node` to `Participant` mapping, the last item still stands and should be solved in a similar fashion.

#### Ignore local publications option

> Currently, there's an `ignore_local_publications` option that can be set when [creting a subscription](https://github.com/ros2/rmw/blob/2250b3eee645d90f9e9d6c96d71ce3aada9944f3/rmw/include/rmw/rmw.h#L517).

Reviewer: @ivanpauno `creting` -> `creating`

That option avoids receiving messages from `Publishers` within the same `Node`.
This wasn't implemented in all the rmw implementations (e.g.: [FastRTPS](https://github.com/ros2/rmw_fastrtps/blob/099f9eed9a0f581447405fbd877c6d3b15f1f26e/rmw_fastrtps_cpp/src/rmw_subscription.cpp#L118)).

For emulating this behavior, messages could be ignored by checking from what `Node` the `Publisher` was created.
This should be possible by querying the state messages, or the local chache were they are organized.

#### Intra process communication

Currently, intra-process communication can be enable disabled in each `Publisher` and `Subscription`.

> There is only one reason for that: intraprocess communication doesn't support all the QoS feature.

Reviewer: @ivanpauno `... support all QoS policies`

Inter process messages from `Publishers` that can also communicate with a `Subscription` using the intra process layer are ignored before handling the callback.
The same problem will happen when having only one `Participant` per context, and it can be solved in the same fashion.

> If in the future our intra process communication support all the QoS policies, we could forgive the possibility of enabling and dissabling it at `Node`, `Publisher`, `Subscription` level.

Reviewer: @ivanpauno `forgive` -> `forbid` ?
